### PR TITLE
Fix and refactor tests

### PR DESCRIPTION
**Pull request by flederwiesel**
Fix test execution by complete refactoring.

Previously, the test scripts have been executed from a rewritten version, which
used `sed` to adjust binary paths. Not only have the tests passed despite the
scripts not being created correctly, this mechanism was also cumbersome and
intransparent.

This refactoring now uses the original test scripts from the source tree.

When building/testing with `cmake`/`ctest`, the (absolute) paths to executables
are being taken from the environment, which is set in CMake via the new
`declare_test_executable()` function.

When executing tests from `make check` (or for manual execution), the current
dir is used. Which is fine, if executed from the cmake build dir.

The path to the files containing the expected data is passed as parameter now.

Furthermore, some optimisations have been done in the bash scripts. Output files
of the test execution are not deleted anymore, which should be OK, if executed
in the build directory. EG_DATA is now also being passed as parameter.

Refactor and add contrib test also, although not clear how this is connected
and how this adds value to contrib...

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -21,6 +21,10 @@ set(shp_LIB_VERSIONINFO "4:0:0")
 set(PROJECT_VERSION
   "${PROJECT_VERSION_MAJOR}.${PROJECT_VERSION_MINOR}.${PROJECT_VERSION_PATCH}")
 
+set(EG_DATA ${CMAKE_CURRENT_BINARY_DIR}/shape_eg_data CACHE STRING "")
+set(EG_DATA_URI http://dl.maptools.org/dl/shapelib/shape_eg_data.zip)
```
Comment (rouault) [inline CMakeLists.txt:25]: I'm not super confident about dl.maptools.org being a long term safe archive. Couldn't we just hava a tests/data/ directory in this repository with the content of this zip ?

Comment (flederwiesel) [reply CMakeLists.txt]: Sure. This way we can also enable `test1.sh` without the hassle of downloading/extracting portably. :)

**Comment by rouault** [comment]
@flederwiesel 
``make distcheck`` fails:
```
~/shapelib/build_ac$ LC_ALL=C make distcheck
make  dist-gzip dist-zip am__post_remove_distdir='@:'
make[1]: Entering directory '/home/even/shapelib/build_ac'
make  distdir-am
make[2]: Entering directory '/home/even/shapelib/build_ac'
make[2]: *** No rule to make target 'tests/stream1.out', needed by 'distdir-am'.  Stop.
make[2]: Leaving directory '/home/even/shapelib/build_ac'
make[1]: *** [Makefile:1247: distdir] Error 2
make[1]: Leaving directory '/home/even/shapelib/build_ac'
make: *** [Makefile:1348: dist] Error 2
```

**Comment by thbeu** [comment]
I believe that all files in tests/shape_eg_data/ also need to be part of the distribution to make the test1.sh passing.

[test-suite.log](https://github.com/OSGeo/shapelib/files/14514454/test-suite.log)

**Comment by rouault** [comment]
@thbeu Not sure if you want to pursue this, but maybe create a new PR then ?

**Comment by thbeu** [comment]
OK, there now is #110 which superseds this PR.

**Pull request closed without merging**